### PR TITLE
Switch to using `importlib` from `pkg_resources`

### LINKED ISSUE #1426: switch to using importlib.metadata from pkgutil
We use pkgutil as one of our plugin mechanisms in python.  A more modern/robust solution is `importlib.metadata`, see the note here:

https://github.com/AcademySoftwareFoundation/OpenTimelineIO/pull/1423#issuecomment-1258732938

After #1423 lands.

### PR DESCRIPTION
**Pull request by mikemahony**
Fixes #1426

As requested in the issue, this replaces `pkg_resources` with `importlib` for plugin detection.

Quick and dirty benchmarks show an improvement in average runtime when accessing plugins:  
image

The `OTIO_DISABLE_PKG_RESOURCE_PLUGINS` env var and its corresponding test have been removed as they're no longer needed.

Less than ideal, `importlib.metadata` has been in flux since `3.7` so there's some code in here to work with old and new and to avoid a deprecation warning.

Existing tests in `test_plugin_detection` continue to provide coverage.

### CODE REVIEW

File: docs/tutorials/otio-env-variables.md
Comment (ssteinbach) [inline docs/tutorials/otio-env-variables.md:25]: Let's keep the environment variable but update the name of it, just in case anyone in the future has problems with this feature (even via the new `importlib` path you're adding.

File: examples/sample_plugin/otio_counter/__init__.py
Comment (ssteinbach) [inline examples/sample_plugin/otio_counter/__init__.py:68]:
```suggestion
    with (
            importlib.resources.path(__name__, "plugin_manifest.json") as p,
            open(p, 'rb') as f
    ):
        return manifest.manifest_from_string(f.read())
```
Style preference for me, consider this optional - I'm not a fan of "\" for line continuation, would rather use ()

File: src/py-opentimelineio/opentimelineio/plugins/manifest.py
Comment (ssteinbach) [inline src/py-opentimelineio/opentimelineio/plugins/manifest.py:289]:
```suggestion
                    with (
                            resources.path(plugin.value, "plugin_manifest.json") as path, 
                            open(path, 'rb') as manifest_stream
                    ):
                        plugin_manifest = core.deserialize_json_from_string(
                            manifest_stream.read().decode("utf-8")
                        )
```
same nitpick

File: src/py-opentimelineio/opentimelineio/plugins/manifest.py
Comment (ssteinbach) [inline src/py-opentimelineio/opentimelineio/plugins/manifest.py:296]: Great idea! could we include the exception string in the log message too?

File: tests/baselines/plugin_module/otio_mockplugin/__init__.py
Comment (ssteinbach) [inline tests/baselines/plugin_module/otio_mockplugin/__init__.py:25]:
```suggestion
    with (
            path(__name__, "unusually_named_plugin_manifest.json") as p,
            open(p, 'rb') as f
    ):
        return manifest.manifest_from_string(f.read())
```

**Review by ssteinbach — DISMISSED** [review]
Looks good, some questions and suggestions!

File: src/py-opentimelineio/opentimelineio/plugins/manifest.py
Comment (JeanChristopheMorinPerso) [inline src/py-opentimelineio/opentimelineio/plugins/manifest.py:15]: This should be added to the project dependencies in the setup.py I think.

File: src/py-opentimelineio/opentimelineio/plugins/manifest.py
Comment (JeanChristopheMorinPerso) [inline src/py-opentimelineio/opentimelineio/plugins/manifest.py:279]: `path` is deprecated. [files](https://docs.python.org/3/library/importlib.resources.html#importlib.resources.files) should be used instead.

Also I believe this except block can simplified:

```
name = plugin_entry_point.__name__
filepath = importlib.resources.files(name) / "plugin_manifest.json"

if filepath.as_posix() in result.source_files:
    continue

try:
    plugin_manifest = core.deserialize_json_from_string(
        filepath.read_text()
    )
    plugin_manifest._update_plugin_source(filepath.as_posix())
    plugin_manifest.source_files.append(filepath.as_posix())
except IOError:
    raise
```

File: tests/test_plugin_detection.py
Comment (JeanChristopheMorinPerso) [inline tests/test_plugin_detection.py:78]: We can't comment on non modified lines, so commenting here: You should modify the `setUp` method to remove the dependency on `pkg_resources`. You can see what I previously did in https://github.com/AcademySoftwareFoundation/OpenTimelineIO/compare/main...JeanChristopheMorinPerso:OpenTimelineIO:importlib?expand=1#diff-3d0c116209043c6578f8bb39f8d5fc32f70fad767498ef5a2e027af4396b5b71.

**Review by JeanChristopheMorinPerso — COMMENTED** [review]
It's funny that you just created this PR. I was about to create one myself soon. I actually already had a [branch](https://github.com/AcademySoftwareFoundation/OpenTimelineIO/compare/main...JeanChristopheMorinPerso:OpenTimelineIO:importlib?expand=1#diff-3d0c116209043c6578f8bb39f8d5fc32f70fad767498ef5a2e027af4396b5b71) 😛

File: docs/tutorials/otio-env-variables.md
Comment (jminor) [reply docs/tutorials/otio-env-variables.md:25]: Yes, please. In some contexts where we know that plugins aren't in use, it will be good to have an option to turn this off & avoid the overhead of looking for them. A toggle is also hypothetically helpful in some security scenarios where you want to limit runtime discovery of untrusted code.

File: src/py-opentimelineio/opentimelineio/plugins/manifest.py
Comment (mikemahony) [reply src/py-opentimelineio/opentimelineio/plugins/manifest.py:279]: Thanks, @JeanChristopheMorinPerso - I love it.

Updating `otio_mockplugin` and `otio_counter` to match and adding support for `3.7`.

File: tests/test_plugin_detection.py
Comment (mikemahony) [reply tests/test_plugin_detection.py:78]: Good call.  Borrowing your code...

File: src/py-opentimelineio/opentimelineio/plugins/manifest.py
Comment (mikemahony) [reply src/py-opentimelineio/opentimelineio/plugins/manifest.py:15]: Thanks for catching that.

File: tests/test_plugin_detection.py
Comment (mikemahony) [reply tests/test_plugin_detection.py:77]: Thank you for providing your branch, @JeanChristopheMorinPerso - it was super helpful.

Comment (JeanChristopheMorinPerso) [reply tests/test_plugin_detection.py:77]: It's a pleasure!

File: src/py-opentimelineio/opentimelineio/plugins/manifest.py
Comment (JeanChristopheMorinPerso) [inline src/py-opentimelineio/opentimelineio/plugins/manifest.py:16]: We could theoretically always rely on `importlib_metadata` until `importlib.metadata` stabilizes and we drop 3.7.

**Review by jminor — APPROVED** [review]
This looks great. Are there any outstanding issues or comments to resolve before merging?

**Comment by JeanChristopheMorinPerso** [comment]
Maybe https://github.com/AcademySoftwareFoundation/OpenTimelineIO/pull/1469#discussion_r1014530096.

File: src/py-opentimelineio/opentimelineio/plugins/manifest.py
Comment (mikemahony) [reply src/py-opentimelineio/opentimelineio/plugins/manifest.py:16]: I think, but can be convinced otherwise, that I'd like to leave it as is.  I like that it's only using the backport when necessary and will otherwise exercise the path moving forward (and gain any performance benefits therein).  Please unresolve if you feel strongly otherwise!

Comment (JeanChristopheMorinPerso) [reply src/py-opentimelineio/opentimelineio/plugins/manifest.py:16]: I find all the try/except cluttering. As for performance, `importlib_metadata` gets all the performance improvements before they land in CPython.

At least, if we move forward with the current state of the PR, we should limit the install of `importlib_metadata` or Pytohn 3.7 only, using a marker like `importlib_metadata; python_version < "3.8"` in `setup.py`.

Comment (jminor) [reply src/py-opentimelineio/opentimelineio/plugins/manifest.py:16]: That seems like a healthy pattern since it ensures that we have lots of proof that the newer module has been working well for a while before we drop the old one. If we rely on the old one, then the drop has more risk. However, @JeanChristopheMorinPerso you mentioned "until importlib.metadata stabilizes" - does that imply that it isn't yet stable? Is there a specific risk there?

Comment (jminor) [reply src/py-opentimelineio/opentimelineio/plugins/manifest.py:16]: Limiting the install to only when needed seems like a good change.

Comment (JeanChristopheMorinPerso) [reply src/py-opentimelineio/opentimelineio/plugins/manifest.py:16]: @jminor `importlib` is stable, so there is no risk. By stabilizing, I meant when they'll stop to change the interface. From what I understand, the interface is "stable" from 3.8 and onward. But I might be mistaken.

Comment (mikemahony) [reply src/py-opentimelineio/opentimelineio/plugins/manifest.py:16]: Good call on the marker, @JeanChristopheMorinPerso.  Added.

**Comment by jminor** [comment]
This is all looking great. @JeanChristopheMorinPerso would you like to push the "Squash and merge" button?

File: setup.py
Comment (JeanChristopheMorinPerso) [inline setup.py:364]: `importlib_metadata==5.0.0` could cause really painful to resolve conflicts to our users. What would be the lowest acceptable version? From https://pypi.org/project/importlib-metadata/#description I see hat 1.4 is equivalent to wwat's i CPython 3.8. SO I guess we could use `importlib_metadata>=1.4; python_version < "3.8"`?

**Review by JeanChristopheMorinPerso — CHANGES_REQUESTED** [review]
One last comment before approving (sorry if I didn't spot that earlier).

File: setup.py
Comment (mikemahony) [reply setup.py]: Another good call.  Updated.

**Review by JeanChristopheMorinPerso — APPROVED** [review]
Thanks a lot for creating that PR (and for following along with the multiple review passes)!

It's pretty exciting to get rid of the runtime dependency on setuptools! (You can't see my face right now, but I can guarantee you that I have a big smile on my face right now).

**Comment by jminor** [comment]
Thanks @mikemahony and @JeanChristopheMorinPerso for bringing this one home. It was great to see y'all making this PR better at each step :)

**Pull request merged**
